api_server/src/lib.rs

```diff
@@ -115,31 +110,35 @@ impl ApiServer {
                 .add(delta_us as usize);
         }
 
-        let http: Http<hyper::Chunk> = Http::new();
-
+        let http = Http::new();
         let f = listener
             .incoming()
-            .for_each(|(stream, _)| {
+            .for_each(move |stream| {
                 // For the sake of clarity: when we use self.efd.clone(), the intent is to
-                // clone the wrapping Rc, not the EventFd itself.
+                // clone the wrapping Arc, not the EventFd itself.
                 let service = ApiServerHttpService::new(
                     self.mmds_info.clone(),
                     self.vmm_shared_info.clone(),
                     self.api_request_sender.clone(),
                     self.efd.clone(),
                 );
-                let connection = http.serve_connection(stream, service);
-                // todo: is spawn() any better/worse than execute()?
-                // We have to adjust the future item and error, to fit spawn()'s definition.
-                handle.spawn(connection.map(|_| ()).map_err(|_| ()));
+                
+                let conn = http
+                    .serve_connection(stream, service)
+                    .map_err(|e| {
+                        error!("server connection error: {}", e);
+                    });
+
+                tokio::spawn(conn);
                 Ok(())
-            }).map_err(Error::Io);
+            }).map_err(|_| ());
 
         // This runs forever, unless an error is returned somewhere within f (but nothing happens
         // for errors which might arise inside the connections we spawn from f, unless we explicitly
         // do something in their future chain). When this returns, ongoing connections will be
         // interrupted, and other futures will not complete, as the event loop stops working.
-        core.run(f)
+        tokio::run(f);
```

Comment:
Given the resource constraints, I would use the current_thead runtime here.

Reply:
This (`tokio::run`) could actually be the reason for the increased memory usage, since this will startup the threadpool runtime, making a thread per CPU, and if the CI has a few, that'd be all the memory. `tokio::runtime::current_thread` won't start any new threads.